core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -1150,22 +1150,24 @@ sealed trait ZIO[-R, +E, +A] extends Serializable with ZIOPlatformSpecific[R, E,
    * in the background.
    */
   final def race[R1 <: R, E1 >: E, A1 >: A](that: ZIO[R1, E1, A1]): ZIO[R1, E1, A1] =
-    ZIO.fiberId
-      .flatMap(parentFiberId =>
-        (self raceWith that)(
-          (exit, right) =>
-            exit.foldM[Any, E1, A1](
-              cause => right.join mapErrorCause (cause && _),
-              a => (right interruptAs parentFiberId) as a
-            ),
-          (exit, left) =>
-            exit.foldM[Any, E1, A1](
-              cause => left.join mapErrorCause (_ && cause),
-              a => (left interruptAs parentFiberId) as a
-            )
-        )
+    ZIO.descriptorWith { descriptor =>
+      val parentFiberId = descriptor.id
+      def maybeDisconnect[R, E, A](zio: ZIO[R, E, A]): ZIO[R, E, A] =
+        if (descriptor.interruptStatus.isInterruptible) zio
+        else zio.uninterruptible.disconnect.interruptible
```

Comment:
Why `zio.uninterruptible.disconnect.interruptible`? It should be sufficient (and more performant) to do:

```
zio.disconnect.interruptible
```

Reply:
@jdegoes So that the underlying `zio` itself remains uninterruptible and can't be sent interrupts to - otherwise it could still be interrupted indirectly by supervision etc.

Reply:
When it's disconnected there's no way to interrupt it aside from having a reference to the fiber.

Reply:
@jdegoes 
It will be interrupted inside `disconnect`:
```scala
  final def disconnect: ZIO[R, E, A] =
    ZIO.uninterruptibleMask(restore =>
      for {
        id    <- ZIO.fiberId
        fiber <- restore(self).forkDaemon
        a     <- restore(fiber.join).onInterrupt(fiber.interruptAs(id).forkDaemon)
      } yield a
    )
```

With `zio.disconnect.interruptible` the following will print "interrupted uninterruptible never":

```scala
ZIO.sleep(5.seconds).race(ZIO.never.onInterrupt(putStrLn("interrupted uninterruptible never")))
```

[Proof](https://scastie.scala-lang.org/VBPVYBOnS2ycowo7s6Sx7Q)

Because the right hand side will be interrupted in `restore(fiber.join).onInterrupt(fiber.interruptAs(id).forkDaemon)`, i.e. omitting `zio.uninterruptible` makes all races interruptible as in RC17 – which I'm probably all for, but doesn't seem like what you intend.

Reply:
Ah right, sorry, I should know that. 😆